[Bug] Rust - Check Database OOM

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/2811a70b-bccf-4ad4-9b94-3386ac6d4196

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
at android.os.AsyncTask$4.done(AsyncTask.java:399)
at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
at java.util.concurrent.FutureTask.run(FutureTask.java:271)
at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:919)
Caused by: java.lang.OutOfMemoryError: Failed to allocate a 169408 byte allocation with 118096 free bytes and 115KB until OOM, target footprint 536870912, growth limit 536870912
at java.lang.StringFactory.newStringFromChars(StringFactory.java:260)
at com.google.protobuf.Utf8$SafeProcessor.decodeUtf8(Utf8.java:20)
at com.google.protobuf.Utf8.decodeUtf8(Utf8.java:2)
at com.google.protobuf.CodedInputStream$ArrayDecoder.readStringRequireUtf8(CodedInputStream.java:3)
at BackendProto.Sqlite$SqlValue.<init>(Sqlite.java:20)
at BackendProto.Sqlite$SqlValue.<init>(Sqlite.java:1)
at BackendProto.Sqlite$SqlValue$1.parsePartialFrom(Sqlite.java:2)
at BackendProto.Sqlite$SqlValue$1.parsePartialFrom(Sqlite.java:1)
at com.google.protobuf.CodedInputStream$ArrayDecoder.readMessage(CodedInputStream.java:14)
at BackendProto.Sqlite$Row.<init>(Sqlite.java:15)
at BackendProto.Sqlite$Row.<init>(Sqlite.java:1)
at BackendProto.Sqlite$Row$1.parsePartialFrom(Sqlite.java:2)
at BackendProto.Sqlite$Row$1.parsePartialFrom(Sqlite.java:1)
at com.google.protobuf.CodedInputStream$ArrayDecoder.readMessage(CodedInputStream.java:14)
at BackendProto.Sqlite$DBResult.<init>(Sqlite.java:15)
at BackendProto.Sqlite$DBResult.<init>(Sqlite.java:1)
at BackendProto.Sqlite$DBResult$1.parsePartialFrom(Sqlite.java:2)
at BackendProto.Sqlite$DBResult$1.parsePartialFrom(Sqlite.java:1)
at com.google.protobuf.CodedInputStream$ArrayDecoder.readMessage(CodedInputStream.java:14)
at BackendProto.Sqlite$DBResponse.<init>(Sqlite.java:16)
at BackendProto.Sqlite$DBResponse.<init>(Sqlite.java:1)
at BackendProto.Sqlite$DBResponse$1.parsePartialFrom(Sqlite.java:2)
at BackendProto.Sqlite$DBResponse$1.parsePartialFrom(Sqlite.java:1)
at com.google.protobuf.AbstractParser.parsePartialFrom(AbstractParser.java:18)
at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:24)
at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:26)
at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:27)
at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:9)
at BackendProto.Sqlite$DBResponse.parseFrom(Sqlite.java:5)
at net.ankiweb.rsdroid.BackendV1Impl.getPage(BackendV1Impl.java:4)
at net.ankiweb.rsdroid.BackendMutex.getPage(BackendMutex.java:2)
at net.ankiweb.rsdroid.database.Session.getPage(Session.java:1)
at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.getNextPage(StreamingProtobufSQLiteCursor.java:3)
at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.moveToNext(StreamingProtobufSQLiteCursor.java:2)
at com.ichi2.libanki.Collection._fieldData(Collection.java:3)
at com.ichi2.libanki.Collection.updateFieldCache(Collection.java:5)
at com.ichi2.libanki.Collection.updateFieldCache(Collection.java:2)
at com.ichi2.libanki.Collection.updateFieldCache(Collection.java:16)
at com.ichi2.libanki.Collection.q(Unknown Source:0)
at com.ichi2.libanki.q.apply(Unknown Source:4)
at com.ichi2.libanki.Collection.lambda$fixIntegrity$1(Collection.java:2)
at com.ichi2.libanki.Collection.j(Unknown Source:0)
at com.ichi2.libanki.g.consume(Unknown Source:8)
at com.ichi2.libanki.Collection.fixIntegrity(Collection.java:27)
at com.ichi2.async.CollectionTask$CheckDatabase.task(CollectionTask.java:5)
at com.ichi2.async.CollectionTask$CheckDatabase.task(CollectionTask.java:1)
at com.ichi2.async.CollectionTask.actualDoInBackground(CollectionTask.java:14)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:2)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:1)
at android.os.AsyncTask$3.call(AsyncTask.java:378)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
... 4 more
```

Initial Thoughts

* Better handling of Rust DB Streaming 
  * Maybe: instead of 1000 rows, determine how to count the size of a protobuf as we stream it into a byte array, and stop at a certain point
   * Or vary the number of rows: since we're getting FieldData, rows can be arbitrarily large
   * Maybe this isn't a blocker if we can "fail forward" and move this to Rust in V2.16.X.
* https://github.com/david-allison-1/Anki-Android-Backend/issues/14 - provide a lambda to the backend which contains "Are we low on/OOM"
* Maybe catch the OOM here and use a fallback which won't fail? This will be slow but better than crashing.

--- a/AnkiDroid/build.gradle
+++ b/AnkiDroid/build.gradle
@@ -248,7 +248,7 @@ dependencies {
     implementation 'com.google.code.gson:gson:2.8.6'
 
     // == Rust conversion (from Anki-Android-Backend on GitHub) ==
-    String backendVersion = "0.1.2" // We want both testing and implementation on the same version
+    String backendVersion = "0.1.4" // We want both testing and implementation on the same version
     // build with ./gradlew rsdroid:assembleRelease
     // In my experience, using `files()` currently requires a reindex operation, which is slow.
     // implementation files("C:\\GitHub\\Rust-Test\\rsdroid\\build\\outputs\\aar\\rsdroid-release.aar")

